File: Demo Programs/VB Demo/MainForm.vb
'*******************************************************************************
'  Copyright 2009-2012 dataweb GmbH
'  This file is part of the NShape framework.
'  NShape is free software: you can redistribute it and/or modify it under the 
'  terms of the GNU General Public License as published by the Free Software 
'  Foundation, either version 3 of the License, or (at your option) any later 
'  version.
'  NShape is distributed in the hope that it will be useful, but WITHOUT ANY
'  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR 
'  A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
'  You should have received a copy of the GNU General Public License along with 
'  NShape. If not, see <http://www.gnu.org/licenses/>.
'*******************************************************************************

Imports Dataweb.NShape
Imports Dataweb.NShape.Advanced
Imports Dataweb.NShape.GeneralShapes


Public Class MainForm
	Dim moveCnt As Integer = 0
	Const stoneSize = 100


	Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
		' Prepare project
		Project.Name = "API DEMO"
		Project.Create()
		Project.AddLibraryByName("Dataweb.NShape.GeneralShapes", False)

		' Prepare diagram
		Dim diagram As New Diagram("Diagram")
		diagram.Width = 3 * stoneSize
		diagram.Height = 3 * stoneSize
		diagram.BackgroundImage = New NamedImage(My.Resources.Background, "Background")
		diagram.BackgroundImageLayout = ImageLayoutMode.Tile

		' Prepare visual styles
		' Create a fill style for the stones
		Dim fillstyle As New FillStyle
		fillstyle.AdditionalColorStyle = Project.Design.ColorStyles.Transparent
		fillstyle.BaseColorStyle = Project.Design.ColorStyles.Black
		fillstyle.FillMode = FillMode.Gradient
		' Create character style for the stone's labels
		Dim characterStyle As New CharacterStyle
		characterStyle.ColorStyle = New ColorStyle("White", Color.White)
		characterStyle.Size = 36
		characterStyle.FontName = "Tahoma"
		characterStyle.Style = Drawing.FontStyle.Bold
		' Create line style for the stone's outlines
		Dim lineStyle As New LineStyle
		lineStyle.ColorStyle = New ColorStyle("White", Color.White)

		' Find shape type for creating the stones
		Dim shapeType As ShapeType
		shapeType = Project.ShapeTypes("RoundedBox")

		Dim r, c As Integer
		For r = 0 To 2
			For c = 0 To 2
				Dim p As Point = New Point((stoneSize / 2) + (c * stoneSize), (stoneSize / 2) + (r * stoneSize))

				If Not (r = 2 And c = 2) Then
					' Create a new stone and set up its visual appearance
					Dim stone As RoundedBox = shapeType.CreateInstance()
					stone.FillStyle = fillstyle
					stone.CharacterStyle = characterStyle
					stone.LineStyle = lineStyle
					' Store the stone's original field assignment in the stone's Tag property
					stone.Tag = GetField(p)
					' Set the stone's size and label
					stone.Width = stoneSize
					stone.Height = stoneSize
					stone.Text = (CInt(stone.Tag) + 1).ToString()
					' Move the stone to its position
					stone.MoveTo(p.X, p.Y)
					' Add stone to the diagram
					diagram.Shapes.Add(stone)
				Else
					' Do nothing
				End If
			Next
		Next
		' Display the diagram
		Display.Diagram = diagram
		' Fit display into form bounds
		ZoomDiagramToForm()
		' Shuffle stone positions
		Shuffle()
	End Sub


	Private Sub Shuffle()
		' Reset move counter
		moveCnt = 0
		' Shuffle stone positions
		Dim Randomizer As New Random(DateTime.Now.Millisecond)
		' Build a list of stones
		Dim stones As New List(Of Shape)(9)
		For i As Integer = 0 To 8
			stones.Add(GetStone(i))
		Next
		For newField As Integer = 0 To 8
			' Pick a random stone out of the list...
			Dim idx As Integer
			idx = Randomizer.Next(1, stones.Count) - 1
			Dim stone As Shape = stones(idx)
			' ... and move it to the current field
			If Not stone Is Nothing Then
				Dim p As Point = GetCoordinates(newField)
				stone.MoveTo(p.X, p.Y)
			End If
			' Remove the stone from the list
			stones.RemoveAt(idx)
		Next
	End Sub


	Private Function GetField(stone As Shape) As Integer
		' Calculate the field of the stone from its coordinates
		If stone Is Nothing Then Return -1
		Dim pos As Point = Point.Empty
		pos.Offset(stone.X, stone.Y)
		Return GetField(pos)
	End Function


	Private Function GetField(position As Point) As Integer
		' Calculate the field of the given coordinates
		Dim row, col As Integer
		col = CInt(Math.Floor(position.X / 100.0F))
		row = CInt(Math.Floor(position.Y / 100.0F))
		Dim field As Integer = col + (row * 3)
		If field < 0 Or field > 8 Then
			Return -1
		Else
			Return field
		End If
	End Function


	Private Function GetCoordinates(field As Integer) As Point
		Dim row, col As Integer
		row = CInt(Math.Floor(field / 3.0F))
		col = field Mod 3
		Dim result As Point
		result.Offset((col * stoneSize) + (stoneSize / 2), (row * stoneSize) + (stoneSize / 2))
		Return result
	End Function


	Private Function GetStone(field As Integer) As Shape
		' Get the stone located in the given field
		Dim pos As Point
		pos = GetCoordinates(field)
		Return GetStone(pos)
	End Function


	Private Function GetStone(position As Point) As Shape
		' Get the stone located at the given coordinates
		For Each shape As Shape In Display.Diagram.Shapes.FindShapes(position.X, position.Y, ControlPointCapabilities.Reference, 0, Nothing)
			If shape.X = position.X And shape.Y = position.Y Then
				Return shape
			End If
		Next
		Return Nothing
	End Function


	Private Sub CheckMoves(ByVal stone As Shape)
		' Get the field of the given stone
		Dim field As Integer = GetField(stone)
		' Get diagram bounds for field validity check
		Dim diagramBounds As Rectangle
		diagramBounds.Location = Point.Empty
		diagramBounds.Width = Display.Diagram.Width
		diagramBounds.Height = Display.Diagram.Height

		' Check for the next free field to move to
		Dim targetField As Integer = -1
		Dim p As Point
		' Check field above
		p = GetCoordinates(field - 3)
		If diagramBounds.Contains(p) And GetStone(p) Is Nothing Then
			targetField = field - 3
		End If
		' Check field left
		p = GetCoordinates(field - 1)
		If diagramBounds.Contains(p) And GetStone(p) Is Nothing Then
			Debug.Assert(targetField < 0)
			targetField = field - 1
		End If
		' Check field right
		p = GetCoordinates(field + 1)
		If diagramBounds.Contains(p) And GetStone(p) Is Nothing Then
			Debug.Assert(targetField < 0)
			targetField = field + 1
		End If
		' Check field below
		p = GetCoordinates(field + 3)
		If diagramBounds.Contains(p) And GetStone(p) Is Nothing Then
			Debug.Assert(targetField < 0)
			targetField = field + 3
		End If

		' If a valid target field was found, move the stone
		If targetField >= 0 Then
			MoveStone(stone, field, targetField)
		End If

		' Check if the game is won after this move
		CheckWon()
	End Sub


	Private Sub CheckWon()
		' Check if all stones are on the fields they are supposed to be
		Dim won As Boolean = True
		For i As Integer = 0 To 7
			Dim stone = GetStone(i)
			If stone Is Nothing Then
				won = False
			Else
				If Not i.Equals(stone.Tag) Then
					won = False
				End If
			End If
			If won = False Then Exit For
		Next
		If won And moveCnt > 0 Then
			MessageBox.Show(String.Format("Game solved with {0} moves. {1}Congratulations!", moveCnt, vbCrLf))
		End If
	End Sub


	Private Sub MoveStone(ByVal stone As Shape, ByVal fromField As Integer, ByVal toField As Integer)
		MoveStone(stone, GetCoordinates(fromField), GetCoordinates(toField))
	End Sub


	Private Sub MoveStone(ByVal stone As Shape, ByVal fromPos As Point, ByVal toPos As Point)
		moveCnt += 1

		Dim stepping As Integer = 0
		Dim steps As Integer = 10

		Dim distanceX As Integer = toPos.X - fromPos.X
		Dim distanceY As Integer = toPos.Y - fromPos.Y

		If distanceX <> 0 Then
			stepping = distanceX / steps
			While stone.X <> toPos.X
				stone.X += stepping
				Application.DoEvents()
			End While
		ElseIf distanceY <> 0 Then
			stepping = distanceY / steps
			While stone.Y <> toPos.Y
				stone.Y += stepping
				Application.DoEvents()
			End While
		End If
		If stone.X <> toPos.X Or stone.Y <> toPos.Y Then
			stone.MoveTo(toPos.X, toPos.Y)
		End If
	End Sub


	Private Sub ZoomDiagramToForm()
		Dim zoom As Single
		If Not Display Is Nothing And Not Display.Diagram Is Nothing Then
			zoom = Math.Min(Display.Bounds.Width / CSng(Display.Diagram.Width), Display.Bounds.Height / CSng(Display.Diagram.Height))
			If (zoom > 0) Then
				Display.ZoomLevel = zoom * 100
			Else
				Display.ZoomLevel = 1
			End If
		End If
	End Sub


	Private Sub Display_ShapeClick(sender As System.Object, e As Dataweb.NShape.Controllers.DiagramPresenterShapeClickEventArgs) Handles Display.ShapeClick
		CheckMoves(e.Shape)
	End Sub


	Private Sub Display_ShapeDoubleClick(sender As System.Object, e As Dataweb.NShape.Controllers.DiagramPresenterShapeClickEventArgs) Handles Display.ShapeDoubleClick
		CheckMoves(e.Shape)
	End Sub


	Private Sub ToolStripButton2_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ToolStripButton2.Click
		Shuffle()
	End Sub


	Private Sub Form1_Resize(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Resize
		ZoomDiagramToForm()
	End Sub


	Private Sub MaximumQualityToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MaximumQualityToolStripMenuItem.Click
		Display.RenderingQualityHighQuality = RenderingQuality.HighQuality
		Display.HighQualityRendering = True
		HighQualityToolStripMenuItem.Checked = False
		MediumQualityToolStripMenuItem.Checked = False
		LowQualityToolStripMenuItem.Checked = False
	End Sub


	Private Sub HighQualityToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles HighQualityToolStripMenuItem.Click
		Display.RenderingQualityHighQuality = RenderingQuality.HighQuality
		Display.HighQualityRendering = True
		MaximumQualityToolStripMenuItem.Checked = False
		MediumQualityToolStripMenuItem.Checked = False
		LowQualityToolStripMenuItem.Checked = False
	End Sub


	Private Sub DefaultQualityToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MediumQualityToolStripMenuItem.Click
		Display.RenderingQualityLowQuality = RenderingQuality.DefaultQuality
		Display.HighQualityRendering = False
		MaximumQualityToolStripMenuItem.Checked = False
		HighQualityToolStripMenuItem.Checked = False
		LowQualityToolStripMenuItem.Checked = False
	End Sub


	Private Sub LowQualityToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles LowQualityToolStripMenuItem.Click
		Display.RenderingQualityLowQuality = RenderingQuality.LowQuality
		Display.HighQualityRendering = False
		MaximumQualityToolStripMenuItem.Checked = False
		HighQualityToolStripMenuItem.Checked = False
		MediumQualityToolStripMenuItem.Checked = False
	End Sub

End Class

File: Demo Programs/Database Designer/MainForm.cs
/******************************************************************************
  Copyright 2009-2012 dataweb GmbH
  This file is part of the NShape framework.
  NShape is free software: you can redistribute it and/or modify it under the 
  terms of the GNU General Public License as published by the Free Software 
  Foundation, either version 3 of the License, or (at your option) any later 
  version.
  NShape is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR 
  A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
  You should have received a copy of the GNU General Public License along with 
  NShape. If not, see <http://www.gnu.org/licenses/>.
******************************************************************************/

using System;
using System.Drawing;
using System.IO;
using System.Reflection;
using System.Windows.Forms;

using Dataweb.NShape;
using Dataweb.NShape.Advanced;
using Dataweb.NShape.GeneralShapes;
using Dataweb.NShape.SoftwareArchitectureShapes;
using Dataweb.NShape.WinFormsUI;


namespace Database_Designer {

	internal enum DockType { 
		None = 0,
		Left = 1, 
		Top = 2, 
		Right = 4, 
		Bottom = 8
	}


	public partial class DatabaseDesignerForm : Form {

		public DatabaseDesignerForm() {
			InitializeComponent();

			toolBoxWindow = new ToolBoxWindow();
			toolBoxWindow.Owner = this;
			toolBoxWindowDockType = DockType.Top ^ DockType.Right ^ DockType.Bottom;

			propertyWindow = new PropertyWindow();
			propertyWindow.Owner = this;
			propertyWindowDockType = DockType.Top ^ DockType.Right ^ DockType.Bottom;
		}

		#region Implementation

		private void MainForm_Load(object sender, EventArgs e) {
			// Connect NShape components
			propertyPresenter.PrimaryPropertyGrid = propertyWindow.PropertyGrid;
			toolSetPresenter.ListView = toolBoxWindow.ListView;
			// Enable loading needed Libraries
			project.AutoLoadLibraries = true;
			// Add library search path
			project.LibrarySearchPaths.Add(Application.StartupPath);
			// Add persistent libraries
			project.AddLibraryByName("Dataweb.NShape.GeneralShapes", false);
			project.AddLibraryByName("Dataweb.NShape.SoftwareArchitectureShapes", false);

			NewProject();
		}


		/// <summary>
		/// Create tools for the shape toolbox and sort them into custom categories
		/// </summary>
		private void CreateTools() {
			toolSetPresenter.ToolSetController.Clear();
			toolSetPresenter.ToolSetController.AddTool(new SelectionTool(), true);

			string category = "Database Entities";
			DatabaseSymbol databaseShape = (DatabaseSymbol)project.ShapeTypes["Database"].CreateInstance();
			databaseShape.Width = 120;
			databaseShape.Height = 120;
			databaseShape.FillStyle = project.Design.FillStyles.Yellow;
			databaseShape.CharacterStyle = project.Design.CharacterStyles.Heading3;
			databaseShape.Text = "Database";
			CreateTemplateAndTool("Database", category, databaseShape);

			EntitySymbol tableShape = (EntitySymbol)project.ShapeTypes["Entity"].CreateInstance();
			tableShape.Width = 100;
			tableShape.Height = 160;
			tableShape.FillStyle = project.Design.FillStyles.Red;
			tableShape.CharacterStyle = project.Design.CharacterStyles.Heading3;
			tableShape.ParagraphStyle = project.Design.ParagraphStyles.Title;
			tableShape.ColumnCharacterStyle = project.Design.CharacterStyles.Caption;
			tableShape.ColumnParagraphStyle = project.Design.ParagraphStyles.Label;
			CreateTemplateAndTool("Entity", category, tableShape);

			RectangularLine line;
			ShapeType relationShapeType = project.ShapeTypes["RectangularLine"];
			line = (RectangularLine)relationShapeType.CreateInstance();
			line.LineStyle = project.Design.LineStyles.Thick;
			CreateTemplateAndTool("Relationship", category, line);

			line = (RectangularLine)relationShapeType.CreateInstance();
			line.LineStyle = project.Design.LineStyles.Thick;
			line.EndCapStyle = project.Design.CapStyles.ClosedArrow;
			CreateTemplateAndTool("1:n Relationship", category, line);

			line = (RectangularLine)relationShapeType.CreateInstance();
			line.LineStyle = project.Design.LineStyles.Thick;
			line.StartCapStyle = project.Design.CapStyles.ClosedArrow;
			line.EndCapStyle = project.Design.CapStyles.ClosedArrow;
			CreateTemplateAndTool("n:m Relationship", category, line);

			CloudSymbol cloudShape = (CloudSymbol)project.ShapeTypes["Cloud"].CreateInstance();
			cloudShape.FillStyle = project.Design.FillStyles.Blue;
			cloudShape.CharacterStyle = project.Design.CharacterStyles.Heading1;
			cloudShape.Width = 300;
			cloudShape.Height = 160;
			cloudShape.Text = "WAN / LAN";
			CreateTemplateAndTool("Cloud", category, cloudShape);

			category = "Description";
			Text text = (Text)project.ShapeTypes["Text"].CreateInstance();
			text.CharacterStyle = project.Design.CharacterStyles.Normal;
			text.Width = 100;
			CreateTemplateAndTool("Text", category, text);

			AnnotationSymbol annotationShape = (AnnotationSymbol)project.ShapeTypes["Annotation"].CreateInstance();
			annotationShape.FillStyle = project.Design.FillStyles.White;
			annotationShape.CharacterStyle = project.Design.CharacterStyles.Caption;
			annotationShape.ParagraphStyle = project.Design.ParagraphStyles.Text;
			annotationShape.Width = 120;
			annotationShape.Height = 120;
			CreateTemplateAndTool("Annotation", category, annotationShape);

			category = "Miscellaneous";
			RoundedBox roundedRectangle = (RoundedBox)project.ShapeTypes["RoundedBox"].CreateInstance();
			roundedRectangle.FillStyle = project.Design.FillStyles.Green;
			roundedRectangle.Width = 120;
			roundedRectangle.Height = 80;
			CreateTemplateAndTool("Box", category, roundedRectangle);

			Ellipse ellipse = (Ellipse)project.ShapeTypes["Ellipse"].CreateInstance();
			ellipse.FillStyle = project.Design.FillStyles.Yellow;
			ellipse.Width = 120;
			ellipse.Height = 80;
			CreateTemplateAndTool("Ellipse", category, ellipse);

			Picture picture = (Picture)project.ShapeTypes["Picture"].CreateInstance();
			picture.FillStyle = project.Design.FillStyles.Transparent;
			picture.Width = 120;
			picture.Height = 120;
			CreateTemplateAndTool("Picture", category, picture);

			ShapeType arcShapeType = project.ShapeTypes["CircularArc"];
			CircularArc arc;
			arc = (CircularArc)arcShapeType.CreateInstance();
			arc.LineStyle = project.Design.LineStyles.Thick;
			CreateTemplateAndTool("Arc", category, arc);

			arc = (CircularArc)arcShapeType.CreateInstance();
			arc.LineStyle = project.Design.LineStyles.Thick;
			arc.EndCapStyle = project.Design.CapStyles.ClosedArrow;
			CreateTemplateAndTool("Bowed Arrow", category, arc);

			arc = (CircularArc)arcShapeType.CreateInstance();
			arc.LineStyle = project.Design.LineStyles.Thick;
			arc.StartCapStyle = project.Design.CapStyles.ClosedArrow;
			arc.EndCapStyle = project.Design.CapStyles.ClosedArrow;
			CreateTemplateAndTool("Bowed Double Arrow", category, arc);
		}


		/// <summary>
		/// Creates a 'TemplateTool' for the shape toolbox and adds it to the specified toolbox category
		/// </summary>
		private void CreateTemplateAndTool(string name, string category, Shape shape) {
			Template template = new Template(name, shape);
			toolSetPresenter.ToolSetController.CreateTemplateTool(template, category);
			project.Repository.InsertAll(template);
		}


		private void AddNewDiagram() {
			// Count diagrams in order to get a decent default name for the new diagram
			int diagramCnt = 0;
			foreach (Diagram d in project.Repository.GetDiagrams())
				diagramCnt++;
			
			// Create the diagram and set default values
			Diagram diagram = new Diagram(string.Format("Diagram {0}", diagramCnt + 1));
			diagram.Width = 1600;
			diagram.Height = 1200;
			// Set background colors and a background image
			diagram.BackgroundGradientColor = Color.WhiteSmoke;
			if (diagramCnt % 2 == 0)
				diagram.BackgroundColor = Color.DarkRed;
			else diagram.BackgroundColor = Color.DarkBlue;
			diagram.BackgroundImage = new NamedImage(Database_Designer.Properties.Resources.NY028_3, "BackgroundImage");
			diagram.BackgroundImageLayout = ImageLayoutMode.Fit;

			// Insert the diagram into the repository (otherwise it won't be saved to file)
			project.Repository.InsertAll(diagram);
			// Display the diagram
			display.Diagram = diagram;

			// Add a menu item to the main menu for displaying this diagram 
			ToolStripMenuItem item = new ToolStripMenuItem();
			item.Text = diagram.Name;
			item.Tag = diagram;
			item.Click += new EventHandler(diagramDropDownItem_Click);
			
			diagramsDropDownButton.DropDown.Items.Add(item);
			diagramsDropDownButton.Text = diagram.Name;
		}


		private void AddLoadedDiagram(Diagram diagram) {
			display.Diagram = diagram;

			// Add a menu item to the main menu for displaying this diagram 
			ToolStripMenuItem item = new ToolStripMenuItem();
			item.Text = diagram.Name;
			item.Tag = diagram;
			item.Click += new EventHandler(diagramDropDownItem_Click);

			diagramsDropDownButton.DropDown.Items.Add(item);
			diagramsDropDownButton.Text = diagram.Name;
		}


		private void ClearLoadedDiagrams() {
			// clear all menu items used for selecting the diagram
			for (int i = diagramsDropDownButton.DropDown.Items.Count - 1; i >= 0; --i) {
				diagramsDropDownButton.DropDown.Items[i].Click -= diagramDropDownItem_Click;
				diagramsDropDownButton.DropDown.Items.RemoveAt(i);
			}
		}


		private void DeleteDiagram(Diagram diagram) {
			// Delete menu item 
			for (int i = diagramsDropDownButton.DropDown.Items.Count - 1; i >= 0; --i) {
				if (diagramsDropDownButton.DropDown.Items[i].Tag == diagram) {
					diagramsDropDownButton.DropDown.Items[i].Click -= new EventHandler(diagramDropDownItem_Click);
					diagramsDropDownButton.DropDown.Items.RemoveAt(i);

					// Delete the diagram including its contents
					project.Repository.DeleteAll(diagram);
					diagram = null;
				}
			}

			// Switch to the previous diagram
			if (display.Diagram == null && diagramsDropDownButton.DropDown.Items.Count > 0)
				display.Diagram = diagramsDropDownButton.DropDown.Items[0].Tag as Diagram;
			else {
				display.Refresh();
				diagramsDropDownButton.Text = "<No Diagrams>";
			}
		}


		private void NewProject(){
			// If a project is already open, close it first.
			if (project.IsOpen) {
				ClearLoadedDiagrams();
				project.Close();
			}

			// Clear project location and file extension and set the default project name
			// This will force the user to select a new location and name for the project when saving.
			xmlStore.DirectoryName = xmlStore.FileExtension = string.Empty;
			project.Name = defaultProjectName;
			project.Create();

			// (Re-)fill the shape toolbox
			CreateTools();

			// Add a new diagram
			AddNewDiagram();
			SetButtonStates();
		}


		private void LoadRepository() {
			OpenFileDialog dlg = new OpenFileDialog();
			dlg.Filter = fileFilter;
			if (dlg.ShowDialog() == DialogResult.OK) {
				if (!string.IsNullOrEmpty(dlg.FileName) && File.Exists(dlg.FileName)) {
					// Close the current project and clear the UI
					project.Close();
					ClearLoadedDiagrams();

					// Set project name, location and file extension
					xmlStore.FileExtension = Path.GetExtension(dlg.FileName);
					xmlStore.DirectoryName = Path.GetDirectoryName(dlg.FileName);
					project.Name = Path.GetFileNameWithoutExtension(dlg.FileName);
					// Load the project from file
					project.Open();

					// As we want a custom toolbox, we have to clear the tools created when loading the 
					// project ans re-fill the toolbox with tools for the existing templates:
					toolSetController.Clear();
					toolSetController.AddTool(new SelectionTool(), true);
					foreach (Template template in project.Repository.GetTemplates()) {
						// Set the default category for all shapes not handled here
						string category = null;
						if (template.Shape is DatabaseSymbol
							|| template.Shape is EntitySymbol
							|| template.Shape is Polyline
							|| template.Shape is CloudSymbol) {
							category = "Database Entities";
						} else if (template.Shape is Text
							|| template.Shape is AnnotationSymbol) {
							category = "Description";
						} else category = "Miscellaneous";

						// Create a tool for the loaded template
						toolSetController.CreateTemplateTool(template, category);
					}

					// And add menu entries for selecting the loaded diagrams
					foreach (Diagram diagram in project.Repository.GetDiagrams())
						AddLoadedDiagram(diagram);
				} else MessageBox.Show(string.Format("File '{0}' does not exist.", dlg.FileName));
			}
		}


		private void SaveRepository() {
			using (SaveFileDialog dlg = new SaveFileDialog()) {
				dlg.CreatePrompt = false;
				dlg.Filter = fileFilter;
				if (!string.IsNullOrEmpty(xmlStore.DirectoryName))
					dlg.InitialDirectory = xmlStore.DirectoryName;
				dlg.FileName = project.Name + (string.IsNullOrEmpty(xmlStore.FileExtension) ? ".nspj" : xmlStore.FileExtension);
				if (dlg.ShowDialog(this) == DialogResult.OK) {
					// Set selected project name (fila name), location and file extension
					project.Name = Path.GetFileNameWithoutExtension(dlg.FileName);
					xmlStore.DirectoryName = Path.GetDirectoryName(dlg.FileName);
					xmlStore.FileExtension = Path.GetExtension(dlg.FileName);
				}
			}
			// Save the project
			if (!string.IsNullOrEmpty(project.Name) && !string.IsNullOrEmpty(xmlStore.DirectoryName))
				project.Repository.SaveChanges();
		}


		private void SetButtonStates() {
			if (display.Diagram != null) {
				cutButton.Enabled = display.DiagramSetController.CanCut(display.Diagram, display.SelectedShapes);
				copyButton.Enabled = display.DiagramSetController.CanCopy(display.SelectedShapes);
				pasteButton.Enabled = display.DiagramSetController.CanPaste(display.Diagram);
				deleteButton.Enabled = display.DiagramSetController.CanDeleteShapes(display.Diagram, display.SelectedShapes);
			}
		}

		#endregion


		#region ToolWindow Docking Imlpementation

		private Rectangle ContentBounds {
			get {
				contentBounds.X = Bounds.X + SystemInformation.Border3DSize.Width;
				contentBounds.Y = Bounds.Y
					+ SystemInformation.CaptionHeight
					+ SystemInformation.Border3DSize.Height
					+ SystemInformation.Border3DSize.Height
					+ SystemInformation.FixedFrameBorderSize.Height
					+ SystemInformation.FixedFrameBorderSize.Height;
				if (toolStripContainer.TopToolStripPanelVisible)
					contentBounds.Y += toolStripContainer.TopToolStripPanel.Height;
				contentBounds.Width = display.DrawBounds.Width;
				contentBounds.Height = display.DrawBounds.Height;
				return contentBounds;
			}
		}


		protected override void OnMove(EventArgs e) {
			base.OnMove(e);
			if (toolBoxWindow.Visible)
				DockToolWindow();
			if (propertyWindow.Visible)
				DockPropertyWindow();
		}


		protected override void OnResize(EventArgs e) {
			base.OnResize(e);
			if (toolBoxWindow != null && toolBoxWindow.Visible)
				DockToolWindow();
			if (toolBoxWindow != null && propertyWindow.Visible)
				DockPropertyWindow();
		}


		private void DockToolWindow() {
			Rectangle contentBounds = Rectangle.Empty;
			contentBounds = ContentBounds;

			try {
				toolBoxWindow.SuspendLayout();
				int toolBoxBottom;
				if (propertyWindow.Visible)
					toolBoxBottom = contentBounds.Top + (contentBounds.Height / 2);
				else
					toolBoxBottom = contentBounds.Bottom;

				if ((toolBoxWindowDockType & DockType.Left) > 0)
					toolBoxWindow.Left = contentBounds.Left;

				if ((toolBoxWindowDockType & DockType.Top) > 0)
					toolBoxWindow.Top = contentBounds.Top + margin;

				if ((toolBoxWindowDockType & DockType.Right) > 0)
					toolBoxWindow.Left = contentBounds.Right - toolBoxWindow.Width - margin;

				if ((toolBoxWindowDockType & DockType.Bottom) > 0)
					toolBoxWindow.Height = toolBoxBottom - toolBoxWindow.Top - margin;
			} finally {
				toolBoxWindow.ResumeLayout();
			}
		}


		private void DockPropertyWindow() {
			Rectangle contentBounds = Rectangle.Empty;
			contentBounds = ContentBounds;

			try {
				propertyWindow.SuspendLayout();
				int propertyWindowTop;
				if (toolBoxWindow.Visible)
					propertyWindowTop = contentBounds.Bottom - (contentBounds.Height / 2);
				else
					propertyWindowTop = contentBounds.Top;

				if ((propertyWindowDockType & DockType.Left) > 0)
					propertyWindow.Left = contentBounds.Left;
				if ((propertyWindowDockType & DockType.Top) > 0)
					propertyWindow.Top = propertyWindowTop + 10;
				if ((propertyWindowDockType & DockType.Right) > 0)
					propertyWindow.Left = contentBounds.Right - propertyWindow.Width - margin;
				if ((propertyWindowDockType & DockType.Bottom) > 0)
					propertyWindow.Height = contentBounds.Bottom - propertyWindow.Top - margin;
			} finally {
				propertyWindow.ResumeLayout();
			}
		}


		private void DatabaseDesignerForm_Shown(object sender, EventArgs e) {
			toolBoxWindow.Show(this);
			DockToolWindow();
			Activate();
		}

		#endregion


		#region Event Handler Implementations (UI)

		private void display_ShapesSelected(object sender, EventArgs e) {
			// Update UI buttons
			SetButtonStates();
		}


		private void toolBoxButton_Click(object sender, EventArgs e) {
			if (toolBoxButton.Checked) {
				toolBoxWindow.Hide();
			}
			else {
				toolBoxWindow.Show();
				this.Focus();
			}
			DockPropertyWindow();
			DockToolWindow();
			toolBoxButton.Checked = toolBoxWindow.Visible;
		}


		private void propertyWindowButton_Click(object sender, EventArgs e) {
			if (propertyWindowButton.Checked) {
				propertyWindow.Hide();
			}
			else {
				propertyWindow.Show();
				this.Focus();
			}
			DockToolWindow();
			DockPropertyWindow();
			propertyWindowButton.Checked = propertyWindow.Visible;
		}


		private void designButton_Click(object sender, EventArgs e) {
			toolSetController.ShowDesignEditor();
		}


		private void diagramButton_Click(object sender, EventArgs e) {
			if (!propertyWindow.Visible) propertyWindowButton_Click(this, null);
			propertyController.SetObject(0, display.Diagram);
		}


		private void gridButton_Click(object sender, EventArgs e) {
			display.IsGridVisible = gridButton.Checked;
		}

		
		private void zoomInButton_Click(object sender, EventArgs e) {
			if (display.ZoomLevel < 10000)
				display.ZoomLevel += zoomStep;
		}


		private void zoomOutButton_Click(object sender, EventArgs e) {
			if (display.ZoomLevel > zoomStep)
				display.ZoomLevel -= zoomStep;
		}

		
		private void cutButton_Click(object sender, EventArgs e) {
			display.Cut(false);
		}


		private void copyButton_Click(object sender, EventArgs e) {
			display.Copy(false);
		}


		private void pasteButton_Click(object sender, EventArgs e) {
			display.Paste(display.GridSize, display.GridSize);
		}


		private void deleteButton_Click(object sender, EventArgs e) {
			display.Delete(false);
		}

		
		private void deleteDiagramButton_Click(object sender, EventArgs e) {
			DeleteDiagram(display.Diagram);
		}
		
		
		private void diagramDropDownItem_Click(object sender, EventArgs e) {
			display.UnselectAll();
			display.Diagram = ((ToolStripMenuItem)sender).Tag as Diagram;
			diagramsDropDownButton.Text = display.Diagram.Name;
		}


		private void addDiagramButton_Click(object sender, EventArgs e) {
			AddNewDiagram();
		}


		private void clearButton_Click(object sender, EventArgs e) {
			NewProject();
		}

		
		private void saveButton_Click(object sender, EventArgs e) {
			SaveRepository();
		}


		private void openButton_Click(object sender, EventArgs e) {
			LoadRepository();
		}

		#endregion


		#region NShape component's event handler implementations

		private void display_ZoomChanged(object sender, EventArgs e) {
			zoomLabel.Text = string.Format("{0} %", display.ZoomLevel);
		}


		private void toolSetController_TemplateEditorSelected(object sender, Dataweb.NShape.Controllers.TemplateEditorEventArgs e) {
			TemplateEditorDialog dlg = new TemplateEditorDialog(e.Project, e.Template);
			dlg.Show(this);
		}


		private void toolSetController_DesignEditorSelected(object sender, EventArgs e) {
			DesignEditorDialog dlg = new DesignEditorDialog(project);
			dlg.Show(this);
		}

		#endregion


		#region Fields

		private ToolBoxWindow toolBoxWindow;
		private PropertyWindow propertyWindow;
		private Rectangle contentBounds = Rectangle.Empty;
		private const int margin = 10;
		private const int zoomStep = 5;
		private const string fileFilter = "NShape XML Repository Files|*.nspj;*.xml|All Files|*.*";
		private const string defaultProjectName = "Project 1";

		private DockType toolBoxWindowDockType = 0;
		private DockType propertyWindowDockType = 0;

		#endregion
	}

}
